docs/remote-snapshotter.md

```diff
@@ -0,0 +1,127 @@
+
+Some remote snapshotters requires snapshotter-specific information through `Pull` API.
+The information will be used in various ways including searching snapshot contents from remote store.
+One of the example snapshotters that requires these information is stargz snapshotter.
+It requires the image reference name and layer digests, etc. for searching layer contents from registries.
+
+The containerd client supports two ways to pass snapshotter-specific information to the underlying snapshotter.
+
+### Using snapshotter's `WithLabels` option
+
+Snapshotter-specific information can be passed as snapshot labels using snapshotter option `WithLabels`.
+Specified labels will be passed to the underlying snapshotter every time the containerd client queries a remote snapshot.
+This is useful if the values of these labels are determined statically regardless of the snapshots.
+Labels must be prefixed by `containerd.io/snapshot/`.
+
+```go
+import "github.com/containerd/containerd/snapshots"
+
+image, err := client.Pull(ctx, ref,
+	containerd.WithPullUnpack,
+	containerd.WithPullSnapshotter(
+		"my-remote-snapshotter",
+		snapshots.WithLabels(map[string]string{
+			"containerd.io/snapshot/reference": ref,
+		}),
+	),
+)
+```
+
+### Using the containerd client's `WithImageHandlerWrapper` option
+
+Snapshotter-specific information can also be passed using an image handler wrapper.
+This is useful when the information varies depending on the snapshot.
+
+Every time the containerd client queries remote snapshot, it passes `Annotations` appended to the targeting layer descriptor (means the layer descriptor that will be pulled and unpacked for preparing that snapshot) to the underlying snapshotter.
+The values of annotations can be dynamically added and modified in the handler wrapper.
+Note that annotations must be prefixed by `containerd.io/snapshot/`.
+[CRI plugin](https://github.com/containerd/cri/blob/09d6426f33cac217528158ddc6d254ca7d597a7b/pkg/server/image_pull.go#L127) and [stargz snapshotter](https://github.com/containerd/stargz-snapshotter/blob/875ec333403a885f5b6e5b64c94ec4dc713e0596/cmd/ctr-remote/commands/rpull.go#L97) levelage this method.
+
+```go
+import "github.com/ktock/snapshotter/handler"
+
+if _, err := client.Pull(ctx, ref,
+	containerd.WithPullUnpack,
+	containerd.WithPullSnapshotter("my-remote-snapshotter"),
+	containerd.WithImageHandlerWrapper(handler.Wrapper(ref)),
+)
+```
+
+## Snapshotter APIs for querying remote snapshots
+
+The containerd client queries remote snapshots to the underlying remote snapshotter using snapshotter APIs.
+This section describes the high-level overview of how snapshotter APIs are used for remote snapshots functionality, with some piece of pseudo-codes that describe the simplified logic implemented in the containerd client.
+For more details, see [`unpacker.go`](/unpacker.go) that implements this logic.
+
+During image pull, the containerd client calls `Prepare` API with label `containerd.io/snapshot.ref`.
```

Comment:
`with the label`

We might also want to mention that `containerd.io/snapshot.ref` is a containerd-defined label, while `containerd.io/snapshot/` is just a label prefix for user-defined (or plugin-defined) labels. Maybe mention user-defined above when you introduce that prefix.